### PR TITLE
new(userspace/libsinsp): support regular expression operator in sinsp filters

### PR DESCRIPTION
**Pull request by jasondellaluce**
**What type of PR is this?**

/kind feature

**Any specific area of the project related to this PR?**

/area libsinsp

/area tests

**Does this PR require a change in the driver versions?**

**What this PR does / why we need it**:

This has been requested by the community since a long time (~6 years) both publicly and in private channels. The main concern of supporting regular expressions is the potential performance impact on high events throughput such as the one of syscalls. However, I consider the project mature enough and in need of serving a multitude of different use cases, so I'm proposing to support this feature and empowering the user with the choice and wisdom on how best using it.

The contribution implements a new `regex` operator that can be used with string fields such as `fd.name regex [a-z]*/proc/[0-9]+/cmdline` or similars. The supported standard is the one of https://github.com/google/re2, on which we base our regex implementation.

This has also the benefit of making performance predictable, as RE2 does not support backtracking and full regex support (from my understanding). On top of that, I added few filter compilation heuristic warnings/checks (with tests) to inform users in cases where regular expression operators are used but not truly needed.

It is crucial that this operator is used with parsimony and only where strictly needed. From my personal benchmarks, the `regex` operator is 5x to 10x than a simple equality comparison for simple examples. As such, existing more trivial operators are always preferable in case they can implement the same check logic.

**Which issue(s) this PR fixes**:

Fixes:
- https://github.com/falcosecurity/falco/issues/438

**Special notes for your reviewer**:

/milestone 0.18.0

**Does this PR introduce a user-facing change?**:

```release-note
new(userpsac/libsinsp): support regular expression operator in sinsp filters
```

### CODE REVIEW

**Comment by incertum** [comment]
Very nice, on that note: Did we ever compare our simple string comparison approach with re2 simple string comparison (no regex)? Just to check if our simple string comparisons have the best performance since re2 is a pretty strong package. WDYT?

**Comment by jasondellaluce** [comment]
@incertum from my experiments, our simple comparisons are generally faster by at least 5x. This is an indicator that the best practice is to use regex comparisons only where regex are the only way of implementing the check

**Comment by incertum** [comment]
> @incertum from my experiments, our simple comparisons are generally faster by at least 5x. This is an indicator that the best practice is to use regex comparisons only where regex are the only way of implementing the check


Great, thanks for sharing!

File: userspace/libsinsp/test/filter_compiler.ut.cpp [hunk 1]
```diff
@@ -681,3 +756,29 @@ TEST_F(sinsp_with_test_input, filter_transformers_wrong_input_type)
...
+TEST_F(sinsp_with_test_input, filter_regex_operator_evaluation)
+{
+	// Basic case just to assert that the basic setup works
+	add_default_init_thread();
+	open_inspector();
+
+	auto evt = generate_getcwd_failed_entry_event();
+
+	// legit use case with a string
+	EXPECT_TRUE(evaluate_filter_str(&m_inspector, "evt.source regex '^[s]{1}ysca[l]{2}$'", evt));
```
Comment (LucaGuerra) [inline userspace/libsinsp/test/filter_compiler.ut.cpp:769]: Note that in https://github.com/falcosecurity/libs/pull/1911 I'm refactoring these kind of tests to be like:

```cpp
EXPECT_TRUE(eval_filter(evt, "evt.source regex '^[s]{1}ysca[l]{2}$'"));
```

The change to these will be very quick to do.

Comment (FedeDP) [reply userspace/libsinsp/test/filter_compiler.ut.cpp:769]: Can we make the changes since #1911 was merged?
This LGTM then :)

Comment (jasondellaluce) [reply userspace/libsinsp/test/filter_compiler.ut.cpp:769]: I apologize for having lost these comments entirely! Let me rebase and push the changes as suggested.

**Review by FedeDP — APPROVED** [review]
/approve

**Review by FedeDP — APPROVED** [review]
/approve

**Review by incertum — APPROVED** [review]
/approve

**Pull request merged**